oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,406 @@
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private String httpMethod;
+    private String url;
+    private String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      if (additionalHeaders.containsKey("x-amz-date")) {
+        try {
+          String xAmzDate = additionalHeaders.get("x-amz-date");
+          new SimpleDateFormat(AwsDates.X_AMZ_DATE_FORMAT).parse(xAmzDate);
+        } catch (ParseException e) {
+          throw new IllegalArgumentException("The provided x-amz-date header value is invalid.", e);
+        }
+      }
+
+      if (additionalHeaders.containsKey("date")) {
+        try {
+          String customDate = additionalHeaders.get("date");
+          new SimpleDateFormat(AwsDates.CUSTOM_DATE_FORMAT).parse(customDate);
+        } catch (ParseException e) {
+          throw new IllegalArgumentException("The provided date header value is invalid.", e);
+        }
+      }
+
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials, httpMethod, url, region, requestPayload, additionalHeaders);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
+
+    private String originalDate;
+    private String amzDate;
+    private String formattedDate;
+
+    private AwsDates(String originalDate, String amzDate, String formattedDate) {
+      this.originalDate = checkNotNull(originalDate);
+      this.amzDate = checkNotNull(amzDate);
+      this.formattedDate = checkNotNull(formattedDate);
+    }
+
+    static AwsDates fromXAmzDate(String xAmzDate) {
+      return new AwsDates(xAmzDate, xAmzDate, xAmzDate.substring(0, 8));
+    }
+
+    static AwsDates fromDateHeader(String date) {
+      DateFormat dateFormat = new SimpleDateFormat(X_AMZ_DATE_FORMAT);
+      dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+      String xAmzDate = null;
+      try {
+        Date inputDate = new SimpleDateFormat(CUSTOM_DATE_FORMAT).parse(date);
+        xAmzDate = dateFormat.format(inputDate);
+      } catch (ParseException e) {
+        // Date is validated at construction.
+      }
+      return new AwsDates(date, xAmzDate, xAmzDate.substring(0, 8));
```

Comment:
This looks like it's going to NPE on a parse error: `xAmzDate` won't be set if the `parse(date)` throws a `ParseException`